### PR TITLE
Fixes #36885 - Add Clevis/Tang disk encryption template

### PR DESCRIPTION
**Pull request by jloeser**
For disk encryption Clevis/Tang is often used. This commit introduces a Kickstart partition template taking care of disk encryption and a snippet responsible for binding the LUKS device via Clevis to a given Tang server.

The default partition template encrypts the disk with a passphrase which can be provided via `disk_enc_passphrase` host parameter. If no host parameter is provided, the default passphrase is 'linux'.

If, in addition, `disk_enc_tang_servers` host parameter is provided (can be one address as string or multiple addresses as array), the LUKS device will be bind to these Tang servers using Clevis. In this case, the passphrase will be removed.

This commit targets mainly all operating systems of the Red Hat family, however the snippet can also be used for Ubuntu operating system.

### CODE REVIEW

**Comment by jloeser** [comment]
https://projects.theforeman.org/issues/36885

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Remove the leading whitepaces, so that it is similar to the if/elsif/end above?

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: 'if" missing?

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Is something else necessary for other OS, too?

Comment (jloeser) [reply app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: I don't understand the question.

Comment (sbernhard) [reply app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: There are two special cases implemented for Redhat and Ubuntu. I wonder, if something else need to be done for Alma, Rocky, Debian, etc?

Comment (jloeser) [reply app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: At the moment only the special cases are tested. There might be more operating systems that could use this snippet (with updated OS conditions), but this was not tested yet. For SUSE for example, this snippet would also work (technically) but the test system wasn't able to decrypt and boot up successfully. Debugging these issues on all operating systems is quite expensive and can be done in future.

Comment (sbernhard) [reply app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: If this was not tested for other OS, then we should maybe just specifiy the related OS in 
https://github.com/theforeman/foreman/pull/9887/files#diff-74ce7a8de62f4fd3cd241aed74d88610178ae49c381d7b7cc8b3deb2908c3406R6

**Comment by jloeser** [comment]
Hey @jhoblitt, I see in the PR that there were some activities done by you. Is there anything missing or I can do to accelerate acceptance of this PR?

**Comment by jhoblitt** [comment]
@jloeser this PR is working for me and would very much like to see it merged. 

/cc @ekohl

**Review by sbernhard — APPROVED** [review]
I would really like to have this in foreman.

**Comment by sbernhard** [comment]
[test foreman]

File: app/views/unattended/provisioning_templates/provision/kickstart_default.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/provision/kickstart_default.erb]: I think, it would be better to only load this snippet if disc enc should be used. So something you did below 'unless tang_server_list.blank? %>' 

Use '<%= -%>" (the minus before %>) which prevents adding useless empty lines.

File: app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb
Comment (sbernhard) [inline app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb]: -%> for preventing empty lines.

File: app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb
Comment (sbernhard) [inline app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb]: -%> for preventing empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Please re-use pkg_manager snippet like in https://github.com/theforeman/foreman/blob/32cedde54ab5b1a0b834d98a5f2730ded982764e/app/views/unattended/provisioning_templates/snippet/redhat_register.erb#L104

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Please re-use pkg_manager snippet like in https://github.com/theforeman/foreman/blob/32cedde54ab5b1a0b834d98a5f2730ded982764e/app/views/unattended/provisioning_templates/snippet/redhat_register.erb#L104

This will reduce the code as it will automatically select the correct pkg manager.

**Review by sbernhard — CHANGES_REQUESTED** [review]
Somre more changes are required before we should merge this.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use -%> to prevent empty lines.

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Use $PKG_MANAGER_INSTALL

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: Would this work for Debian / SuSE, too?

**Review by sbernhard — CHANGES_REQUESTED** [review]

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (jloeser) [reply app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: For SUSE it doesn't work out-of-the-box and the host fails to boot up (probably something is missing in initrd). Debian was not tested.

File: app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb]: 2 whitespaces after os_major and os_minor ;-)

**Review by sbernhard — DISMISSED** [review]
Thank you very much.

File: app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb
Comment (evgeni) [inline app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb]:
```suggestion
autopart --encrypted --passphrase="<%= host_param('disk_enc_passphrase', 'linux') %>" <%= host_param('autopart_options') %>
```

File: app/views/unattended/partition_tables_templates/preseed_default_autoinstall_encrypted.erb
Comment (evgeni) [inline app/views/unattended/partition_tables_templates/preseed_default_autoinstall_encrypted.erb]:
```suggestion
    password: <%= host_param('disk_enc_passphrase', 'linux') %>
```

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (evgeni) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]:
```suggestion
  passphrase = host_param('disk_enc_passphrase', 'linux')
```

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/Preseed_Autoinstall_cloud-init_user_data.ubuntu_autoinst4dhcp.snap.txt
Comment (evgeni) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/Preseed_Autoinstall_cloud-init_user_data.ubuntu_autoinst4dhcp.snap.txt:36]: these empty lines look odd, especially the the second one.

**Review by evgeni — DISMISSED** [review]

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/Preseed_Autoinstall_cloud-init_user_data.ubuntu_autoinst4dhcp.snap.txt
Comment (jloeser) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/Preseed_Autoinstall_cloud-init_user_data.ubuntu_autoinst4dhcp.snap.txt:36]: Yeah, what should I say: building pretty-formated YAML out of templates with conditional includes/snippets is meh.

If the snippet is inserted, then we definitely need a newline to produce valid YAML/cloud-config. But if not we get the newline anyway because of `<%= indent(2) { snippet ... if ... } %>`.

Alternatively, we could add a blank line at the end of the snippet and use  `<%= ... -%>`. Tell me what's the best way to solve this.

**Comment by evgeni** [comment]
ok to test

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/Preseed_Autoinstall_cloud-init_user_data.ubuntu_autoinst4dhcp.snap.txt
Comment (evgeni) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/Preseed_Autoinstall_cloud-init_user_data.ubuntu_autoinst4dhcp.snap.txt:36]: Yeah. Templating YAML is awful, I agree.
I think I do prefer the `-%>` way, but wouldn't block on this if you find it too ugly :)

**Review by sbernhard — DISMISSED** [review]

File: app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb]: This is ERB, so you use `&&`. That, and prevent empty lines
```suggestion
<%= indent(2) { snippet 'preseed_autoinstall_clevis_tang_wrapper' if host_param('disk_enc_tang_servers') && os_major >= 22 && os_minor >= 3 } -%>
```

File: app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb:72]: Please keep the whitespace trimming at the end.
```suggestion
<%= indent(2) { @host.diskLayout } -%>
```

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb:22]: If the param is not a string, then what is the type?

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb]: This so ERB, so please use `&&` and `||` instead
```suggestion
<% if (@host.operatingsystem.family == 'Redhat' || @host.operatingsystem.name == 'Ubuntu') && unless tang_server_list.blank? -%>
```

File: app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb
Comment (ekohl) [inline app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb:18]: this implies the default password is `linux` if none is given. Is this documented? Would it make sense to not encrypt if no passphrase is given?

**Review by ekohl — COMMENTED** [review]
Reading https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption_security-hardening#configuring-automated-enrollment-of-luks-encrypted-volumes-using-kickstart_configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption this looks correct.

**Comment by dosas** [comment]
Some out of scope comment (for this PR):
I see a lot of comments regarding the style of erb templates. While these comments are valid, it is quite annoying to 1. mention them in review and 2. fix them. Especially when there are tools that can do that automated and save time and iterations: https://github.com/Shopify/erb-lint 

https://projects.theforeman.org/issues/37281

File: app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb
Comment (jloeser) [reply app/views/unattended/provisioning_templates/snippet/disk_enc_clevis_tang.erb:22]: Can also be an array (see https://github.com/theforeman/foreman/pull/9887/commits/0c41804d3774276216a151fd3f06f384fc7908d0).

Documentation is work-in-progress here: https://github.com/theforeman/foreman-documentation/pull/2871

File: app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb
Comment (jloeser) [reply app/views/unattended/provisioning_templates/user_data/preseed_autoinstall_cloud_init.erb]: Regarding empty lines, please have a look at the discussion above* and tell me explicitly which way you guys prefer. As said, avoiding empty lines in ERB requires empty lines in the corresponding snippets.

* https://github.com/theforeman/foreman/pull/9887/files/0c41804d3774276216a151fd3f06f384fc7908d0#diff-02b11c45a8fe92a6ae919a699d0198e01b0f9fca7b3449725e6406ebae836fa3

File: app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb
Comment (jloeser) [reply app/views/unattended/partition_tables_templates/kickstart_default_encrypted.erb:18]: Documentation is work-in-progress here: https://github.com/theforeman/foreman-documentation/pull/2871

> Would it make sense to not encrypt if no passphrase is given?

This would require to always provide a intermediate passphrase via host parameter along with Tang server addresses. 
It's more obviously to the user what he can expect from the partition table when there are "encryption" and "non encryption" templates. But I am dispassionate on this topic.

**Comment by sbernhard** [comment]
[test foreman]

**Comment by sbernhard** [comment]
Are there still some open discussion @evgeni  / @ekohl ?
AFAIK, the only remaining thing is the question: https://github.com/theforeman/foreman/pull/9887#discussion_r1530110984
How should we proceed?

**Review by sbernhard — APPROVED** [review]

**Comment by sbernhard** [comment]
Thank you.

**Pull request merged**
